src/core/signals.cpp

```diff
@@ -3,18 +3,161 @@
 
 #include "common/arch.h"
 #include "common/assert.h"
+#include "common/decoder.h"
 #include "core/signals.h"
 
 #ifdef _WIN32
 #include <windows.h>
 #else
 #include <csignal>
 #ifdef ARCH_X86_64
-#include <Zydis/Decoder.h>
 #include <Zydis/Formatter.h>
+#include <sys/ucontext.h>
 #endif
 #endif
 
+namespace {
+
+#if defined(ARCH_X86_64)
+
+#if defined(_WIN32)
+#define INCREMENT_RIP(ctx, length) ((CONTEXT*)ctx)->Rip += length
+#elif defined(__APPLE__)
+#define INCREMENT_RIP(ctx, length) ((ucontext_t*)ctx)->uc_mcontext->__ss.__rip += length
+#else
+#define INCREMENT_RIP(ctx, length) ((ucontext_t*)ctx)->uc_mcontext.gregs[REG_RIP] += length
+#endif
+
+void* getXmmPointer(void* ctx, u32 index) {
+#if defined(_WIN32)
+#define CASE(index)                                                                                \
+    case index:                                                                                    \
+        return (void*)(&(((CONTEXT*)ctx)->Xmm##index.Low))
+#elif defined(__APPLE__)
+#define CASE(index)                                                                                \
+    case index:                                                                                    \
+        return (void*)(&((ucontext_t*)ctx)->uc_mcontext->__fs.__fpu_xmm##index);
+#else
+#define CASE(index) return (void*)(&((ucontext_t*)ctx)->uc_mcontext.fpregs->_xmm[index].element[0])
+#endif
+    switch (index) {
+        CASE(0);
+        CASE(1);
+        CASE(2);
+        CASE(3);
+        CASE(4);
+        CASE(5);
+        CASE(6);
+        CASE(7);
+        CASE(8);
+        CASE(9);
+        CASE(10);
+        CASE(11);
+        CASE(12);
+        CASE(13);
+        CASE(14);
+        CASE(15);
+    default: {
+        UNREACHABLE_MSG("Invalid XMM register index: {}", index);
+        return nullptr;
+    }
+    }
+#undef CASE
+}
+
+// We need to check, before patching, if there's enough space for a relative jump to the trampoline.
+// If there isn't, the instruction must be handled specially in the illegal instruction handler
+// itself.
+bool shouldNotBePatched(void* code_address) {
+    ZydisDecodedInstruction instruction;
+    ZydisDecodedOperand operands[ZYDIS_MAX_OPERAND_COUNT];
+    const auto status =
+        Common::Decoder::Instance()->decodeInstruction(instruction, operands, code_address);
+    if (!ZYAN_SUCCESS(status)) {
+        LOG_ERROR(Core, "Failed to decode instruction at: {}", fmt::ptr(code_address));
+    }
+
+    if (instruction.length < 5) {
+        // not enough bytes for a relative jump for the trampoline
+        return true;
+    }
+
+    return false;
+}
+
+bool handleIllegalInstruction(void* ctx, void* code_address) {
```

Comment:
This should all be part of the `cpu_patches` handler, this file is meant to be the generic dispatch. Adding a bunch of specific use cases here will clutter things up too much. The utilities for accessing the context in platform-specific ways should stay here though.

Reply:
I'll move the utilities to a separate file since they don't really belong in signals either imo

Reply:
Done, could you give this another read?